lib/workflow/task-graph.js

```diff
@@ -251,21 +251,36 @@ function taskGraphFactory(
         })
         .then(function() {
             return Promise.map(self.definition.tasks, function(taskData) {
-                if (_.has(taskData, 'taskDefinition')) {
-                    self._validateTaskDefinition(taskData.taskDefinition);
-                    self._validateProperties(taskData.taskDefinition, context);
-                    self._validateOptions(taskData.taskDefinition, taskData.label);
-                } else {
+                if (!_.has(taskData, 'taskDefinition')) {
                     return store.getTaskDefinition(taskData.taskName)
-                    .then(function(taskDefinition) {
-                        self._validateTaskDefinition(taskDefinition);
-                        self._validateProperties(taskDefinition, context);
-                        self._validateOptions(taskDefinition, taskData.label);
+                    .then(function(definition) {
+                        return {
+                            taskDefinition: definition,
+                            label: taskData.label
+                        };
                     });
+                } else {
+                    return {
+                        taskDefinition: taskData.taskDefinition,
+                        label: taskData.label
+                    };
                 }
+            })
+            .then(function(tasks) {
+                _.forEach(tasks, function(taskData) {
+                    if (_.has(taskData, 'taskDefinition')) {
+                        self._validateTaskDefinition(taskData.taskDefinition);
+                        self._validateProperties(taskData.taskDefinition, context);
+                        self._validateOptions(taskData.taskDefinition, taskData.label);
+                    } else {
+                        self._validateTaskDefinition(taskData.taskDefinition);
+                        self._validateProperties(taskData.taskDefinition, context);
+                        self._validateOptions(taskData.taskDefinition, taskData.label);
```

Comment:
This `if/else` branches call the same functions with same arguments, so why need the branch?